### PR TITLE
Clamp compression level

### PR DESCRIPTION
**Pull request by animalize**
Clamp compression level where needed.

It should cover all possible sites.
If this code style is not liked, feel free to ask me to modify or submit a new PR.

It's up to you to discuss whether to clamp silently or to return an error.
IMO clamp is safe, suppose the maximum level of zstd v1.4 is 22, the maximum level of zstd v1.5 is 30, then the code written for v1.5 can run on v1.4 without error.

This change seems to fix a wrong parameter type:
```diff
--- a/lib/dictBuilder/zdict.c
+++ b/lib/dictBuilder/zdict.c
@@ -706,7 +705,7 @@ static void ZDICT_flatLit(unsigned* countLit)
 
 #define OFFCODE_MAX 30  /* only applicable to first block */
 static size_t ZDICT_analyzeEntropy(void*  dstBuffer, size_t maxDstSize,
-                                   unsigned compressionLevel,
+                                   int compressionLevel,
                              const void*  srcBuffer, const size_t* fileSizes, unsigned nbFiles,
                              const void* dictBuffer, size_t  dictBufferSize,
                                    unsigned notificationLevel)
```

Issue https://github.com/facebook/zstd/issues/2350

### CODE REVIEW

File: lib/common/zstd_internal.h [hunk 1]
```diff
@@ -342,6 +342,14 @@ typedef enum {
     ZSTD_bm_stable = 1     /* ZSTD_inBuffer/ZSTD_outBuffer is stable */
 } ZSTD_bufferMode_e;
 
+static INLINE_KEYWORD UNUSED_ATTR
+void clampComperssionLevel(int *compressionLevel)
```
Comment (Cyan4973) [inline lib/common/zstd_internal.h:346]: `clampComperssionLevel` => `clampCompressionLevel`

**Comment by Cyan4973** [comment]
Clamping of compression level is supposed already achieved by `ZSTD_cParam_clampBounds(ZSTD_c_compressionLevel, &value)`.
Why do you want to rewrite it as `clampCompressionLevel(&value)` ?

File: lib/dictBuilder/zdict.c [hunk 2]
```diff
@@ -706,7 +705,7 @@ static void ZDICT_flatLit(unsigned* countLit)
 
 #define OFFCODE_MAX 30  /* only applicable to first block */
 static size_t ZDICT_analyzeEntropy(void*  dstBuffer, size_t maxDstSize,
-                                   unsigned compressionLevel,
+                                   int compressionLevel,
```
Comment (Cyan4973) [inline lib/dictBuilder/zdict.c:708]: 👍

**Comment by animalize** [comment]
> Clamping of compression level is supposed already achieved by `ZSTD_cParam_clampBounds(ZSTD_c_compressionLevel, &value)`.
> Why do you want to rewrite it as `clampCompressionLevel(&value)` ?

IMO `ZSTD_cParam_clampBounds()` has extra cost.
It calls `ZSTD_cParam_getBounds()`, copies struct value, checks `ZSTD_isError()`, these are unnecessary for compression level.

If you don't like it, I can change it, or feel free to create a new PR. I understand reviewing this PR takes as much work as creating a new one.

File: lib/common/zstd_internal.h (hunk 1 above)
Comment (animalize) [reply lib/common/zstd_internal.h]: Fixed.

**Comment by animalize** [comment]
It seems compression level can be clamped only in `ZSTD_getCParams_internal()` function, other sites don't need to be modified.

https://github.com/facebook/zstd/blob/f62edf0fe9b69408dd3021c29754589ea8dd965a/lib/compress/zstd_compress.c#L4680-L4695

**Review by terrelln — APPROVED** [review]
Thanks for the PR! This looks good to me and will be part of the next release, zstd-1.4.7.

**Comment by animalize** [comment]
I just read an article about avoiding `else` statement as much as possible.
Maybe this patch is better, it's more reasonable for `cp.targetLength`.

```diff
 lib/compress/zstd_compress.c | 7 ++++++-
 1 file changed, 6 insertions(+), 1 deletion(-)

diff --git a/lib/compress/zstd_compress.c b/lib/compress/zstd_compress.c
index b1bb9fa1..4d4bec8e 100644
--- a/lib/compress/zstd_compress.c
+++ b/lib/compress/zstd_compress.c
@@ -4513,7 +4513,12 @@ static ZSTD_compressionParameters ZSTD_getCParams_internal(int compressionLevel,
     if (compressionLevel < 0) row = 0;   /* entry 0 is baseline for fast mode */
     if (compressionLevel > ZSTD_MAX_CLEVEL) row = ZSTD_MAX_CLEVEL;
     {   ZSTD_compressionParameters cp = ZSTD_defaultCParameters[tableID][row];
-        if (compressionLevel < 0) cp.targetLength = (unsigned)(-compressionLevel);   /* acceleration factor */
+        if (compressionLevel < 0) {
+            /* acceleration factor */
+            cp.targetLength = (unsigned)(-compressionLevel);
+            if (cp.targetLength > ZSTD_TARGETLENGTH_MAX)
+                cp.targetLength = ZSTD_TARGETLENGTH_MAX;
+        }
         /* refine parameters based on srcSize & dictSize */
         return ZSTD_adjustCParams_internal(cp, srcSizeHint, dictSize, mode);
     }
```


And changes in `/lib/dictBuilder/zdict.c`, can be done in a separate PR:
```diff
--- a/lib/dictBuilder/zdict.c
+++ b/lib/dictBuilder/zdict.c
@@ -706,7 +705,7 @@ static void ZDICT_flatLit(unsigned* countLit)
 
 #define OFFCODE_MAX 30  /* only applicable to first block */
 static size_t ZDICT_analyzeEntropy(void*  dstBuffer, size_t maxDstSize,
-                                   unsigned compressionLevel,
+                                   int compressionLevel,
                              const void*  srcBuffer, const size_t* fileSizes, unsigned nbFiles,
                              const void* dictBuffer, size_t  dictBufferSize,
                                    unsigned notificationLevel)
```

Reconsider whether `g_compressionLevel_default` variable can be replaced by `ZSTD_CLEVEL_DEFAULT`:
https://github.com/facebook/zstd/blob/v1.4.5/lib/dictBuilder/zdict.c#L65

**Comment by terrelln** [comment]
I think the patch is clear as-is. Though a separate PR for the two `dictBuilder` changes you mentioned would be very welcome!

**Pull request merged**